src/libraries/System.Private.Uri/src/System/UriHelper.cs

```diff
@@ -742,41 +746,68 @@ public static int Compress(Span<char> span, bool convertPathSlashes, bool canoni
                 {
                     if (removeSegments != 0)
                     {
-                        --removeSegments;
-
-                        span.Slice(lastSlash + 1).CopyTo(span.Slice(i + 1));
-                        span = span.Slice(0, span.Length - (lastSlash - i));
+                        removeSegments--;
+                        removedSegments.Append((i + 1, lastSlash - i));
                     }
+
                     lastSlash = i;
                 }
             }
 
-            if (span.Length != 0 && canonicalizeAsFilePath)
+            if (canonicalizeAsFilePath)
             {
                 if (slashCount <= 1)
                 {
                     if (removeSegments != 0 && span[0] != '/')
                     {
-                        //remove first not rooted segment
-                        lastSlash++;
-                        span.Slice(lastSlash).CopyTo(span);
-                        return span.Length - lastSlash;
+                        // remove first not rooted segment
+                        removedSegments.Append((0, lastSlash + 1));
                     }
                     else if (dotCount != 0)
                     {
                         // If final string starts with a segment looking like .[...]/ or .[...]<eos>
                         // then we remove this first segment
                         if (lastSlash == dotCount || (lastSlash == 0 && dotCount == span.Length))
                         {
-                            dotCount += lastSlash == 0 ? 0 : 1;
-                            span.Slice(dotCount).CopyTo(span);
-                            return span.Length - dotCount;
+                            removedSegments.Append((0, dotCount + (lastSlash == 0 ? 0 : 1)));
                         }
                     }
                 }
             }
 
-            return span.Length;
+            if (removedSegments.Length == 0)
+            {
+                return span.Length;
+            }
+
+            // Merge any remaining segments
+            int writeOffset = 0;
+            int readOffset = 0;
+
+            for (int i = removedSegments.Length - 1; i >= 0; i--)
+            {
+                (int start, int length) = removedSegments[i];
+
+                Debug.Assert(start >= readOffset && length > 0 && start + length <= span.Length);
+
+                if (readOffset != start)
+                {
+                    int segmentLength = start - readOffset;
+                    span.Slice(readOffset, segmentLength).CopyTo(span.Slice(writeOffset));
```

Comment:
NIT: I am not sure how well is Span.CopyTo optimized for noop copy. Please consider to skip it `if (readoffset == writeOffset) for performance reasons.

Reply:
If the source and destination overlap perfectly, the copy operation will be skipped: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.ByteMemOps.cs,230

Reply:
The read and write offset can only ever be the same for the one section before the first removed segment, so we'd avoid at most one call.

Since it only concerns the first copy, I did effectively the same thing but avoiding the extra branches in daad5170d4b92746f81587a2d1fee1de3a21ee79 - we can just ignore anything before the start of the first removed segment.